Support servant-0.9

The lib is ok, but an example needs adjusting

```
example/Main.hs:21:17: error:
    Module ‘Servant’ does not export ‘FromFormUrlEncoded(..)’
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,12 +4,25 @@ sudo: false
 
 matrix:
   include:
-    - env: CABALVER=1.18 GHCVER=7.8.4
-      addons: {apt: {packages: [cabal-install-1.18,ghc-7.8.4], sources: [hvr-ghc]}}
-    - env: CABALVER=1.22 GHCVER=7.10.3
-      addons: {apt: {packages: [cabal-install-1.22,ghc-7.10.3], sources: [hvr-ghc]}}
+    # Unrestricted
     - env: CABALVER=1.24 GHCVER=8.0.1
       addons: {apt: {packages: [cabal-install-1.24,ghc-8.0.1], sources: [hvr-ghc]}}
+    - env: CABALVER=1.22 GHCVER=7.10.3
+      addons: {apt: {packages: [cabal-install-1.22,ghc-7.10.3], sources: [hvr-ghc]}}
+    - env: CABALVER=1.18 GHCVER=7.8.4
+      addons: {apt: {packages: [cabal-install-1.18,ghc-7.8.4], sources: [hvr-ghc]}}
+    # Constrained
+    - env: CABALVER=1.24 GHCVER=8.0.1 CABALCONFIG=servant=0.9
+      addons: {apt: {packages: [cabal-install-1.24,ghc-8.0.1], sources: [hvr-ghc]}}
+    - env: CABALVER=1.24 GHCVER=8.0.1 CABALCONFIG=servant=0.8
+      addons: {apt: {packages: [cabal-install-1.24,ghc-8.0.1], sources: [hvr-ghc]}}
+    - env: CABALVER=1.22 GHCVER=7.10.3 CABALCONFIG=servant-0.7
+      addons: {apt: {packages: [cabal-install-1.22,ghc-7.10.3], sources: [hvr-ghc]}}
+    - env: CABALVER=1.22 GHCVER=7.10.3 CABALCONFIG=servant-0.6
+      addons: {apt: {packages: [cabal-install-1.22,ghc-7.10.3], sources: [hvr-ghc]}}
+    - env: CABALVER=1.22 GHCVER=7.10.3 CABALCONFIG=servant-0.5
+      addons: {apt: {packages: [cabal-install-1.22,ghc-7.10.3], sources: [hvr-ghc]}}
+    # HEAD
     - env: CABALVER=head GHCVER=head
       addons: {apt: {packages: [cabal-install-head,ghc-head], sources: [hvr-ghc]}}
 
@@ -23,6 +36,7 @@ install:
  - cabal --version
  - echo "$(ghc --version) [$(ghc --print-project-git-commit-id 2> /dev/null || echo '?')]"
  - travis_retry cabal update
+ - if [ ! -z "$CABALCONFIG" ]; then cp .travis/${CABALCONFIG}-cabal.config cabal.config; fi
  - cabal install --only-dependencies --enable-tests
 
 script:

--- a/.travis/servant-0.5-cabal.config
+++ b/.travis/servant-0.5-cabal.config
@@ -1,0 +1,1 @@
+constraints: servant ==0.5.*

--- a/.travis/servant-0.6-cabal.config
+++ b/.travis/servant-0.6-cabal.config
@@ -1,0 +1,1 @@
+constraints: servant ==0.6.*

--- a/.travis/servant-0.7-cabal.config
+++ b/.travis/servant-0.7-cabal.config
@@ -1,0 +1,1 @@
+constraints: servant ==0.7.*

--- a/.travis/servant-0.8-cabal.config
+++ b/.travis/servant-0.8-cabal.config
@@ -1,0 +1,1 @@
+constraints: servant ==0.8.*

--- a/.travis/servant-0.9-cabal.config
+++ b/.travis/servant-0.9-cabal.config
@@ -1,0 +1,1 @@
+constraints: servant ==0.9.*

--- a/example/Main.hs
+++ b/example/Main.hs
@@ -11,20 +11,26 @@ module Main (main) where
 
 import Control.Monad
 import Crypto.Random (drgNew)
-import Data.ByteString (ByteString)
 import Data.Default
 import Data.List (find)
 import Data.Serialize (Serialize)
 import GHC.Generics
 import Network.Wai (Application, Request)
 import Network.Wai.Handler.Warp (run)
-import Servant (FromFormUrlEncoded(..), FormUrlEncoded, (:<|>)(..), (:>), ReqBody)
+#if MIN_VERSION_servant (0,9,0)
+import Web.FormUrlEncoded (FromForm(..), lookupUnique)
+#else
+import Servant (FromFormUrlEncoded(..))
+#endif
+import Servant ((:<|>)(..), (:>), ReqBody, FormUrlEncoded)
 import Servant (Post, Headers, Header, AuthProtect, Get, Server, Proxy)
 import Servant (addHeader, serveWithContext, Proxy(..), Context(..))
 import Servant.HTML.Blaze
 import Servant.Server.Experimental.Auth (AuthHandler)
 import Servant.Server.Experimental.Auth.Cookie
 import Text.Blaze.Html5 ((!), Markup)
+import System.Exit (exitSuccess)
+import System.Environment (getArgs)
 import qualified Data.Text as T
 import qualified Text.Blaze.Html5 as H
 import qualified Text.Blaze.Html5.Attributes as A
@@ -48,6 +54,15 @@ data LoginForm = LoginForm
   , lfPassword :: String
   } deriving (Eq, Show)
 
+#if MIN_VERSION_servant (0,9,0)
+instance FromForm LoginForm where
+  fromForm f = do
+    username <- fmap T.unpack $ lookupUnique "username" f
+    password <- fmap T.unpack $ lookupUnique "password" f
+    return LoginForm
+      { lfUsername = username
+      , lfPassword = password }
+#else
 instance FromFormUrlEncoded LoginForm where
   fromFormUrlEncoded d = do
     username <- case lookup "username" d of
@@ -59,6 +74,7 @@ instance FromFormUrlEncoded LoginForm where
     return LoginForm
       { lfUsername = username
       , lfPassword = password }
+#endif
 
 usersDB :: [Account]
 usersDB =
@@ -74,7 +90,7 @@ type ExampleAPI =
        Get '[HTML] Markup
   :<|> "login" :> Get '[HTML] Markup
   :<|> "login" :> ReqBody '[FormUrlEncoded] LoginForm
-       :> Post '[HTML] (Headers '[Header "set-cookie" ByteString] Markup)
+       :> Post '[HTML] (Headers '[Header "set-cookie" EncryptedSession] Markup)
   :<|> "private" :> AuthProtect "cookie-auth" :> Get '[HTML] Markup
 
 server :: AuthCookieSettings -> RandomSource -> ServerKey -> Server ExampleAPI
@@ -88,7 +104,7 @@ server settings rs sk = serveHome
 
   serveLoginPost LoginForm {..} =
     case userLookup lfUsername lfPassword usersDB of
-      Nothing   -> return $ addHeader "" (loginPage False)
+      Nothing   -> return $ addHeader emptyEncryptedSession (loginPage False)
       Just uid -> addSession
         settings -- the settings
         rs       -- random source
@@ -106,6 +122,10 @@ app settings rs sk = serveWithContext
 
 main :: IO ()
 main = do
+  args <- getArgs
+  when (args /= ["run"]) $ do
+      putStrLn "Use './example run' to run an example"
+      exitSuccess
   rs <- mkRandomSource drgNew 1000
   sk <- mkServerKey 16 Nothing
   run 8080 (app def rs sk)

--- a/example/Main.hs
+++ b/example/Main.hs
@@ -9,6 +9,9 @@
 
 module Main (main) where
 
+import Prelude ()
+import Prelude.Compat
+
 import Control.Monad
 import Crypto.Random (drgNew)
 import Data.Default
@@ -34,10 +37,6 @@ import System.Environment (getArgs)
 import qualified Data.Text as T
 import qualified Text.Blaze.Html5 as H
 import qualified Text.Blaze.Html5.Attributes as A
-
-#if !MIN_VERSION_base(4,8,0)
-import Control.Applicative
-#endif
 
 data Account = Account
   { accUid       :: Int

--- a/servant-auth-cookie.cabal
+++ b/servant-auth-cookie.cabal
@@ -70,6 +70,7 @@ test-suite example
   type:                exitcode-stdio-1.0
   main-is:             Main.hs
   build-depends: base           >= 4.7  && < 5.0
+               , base-compat    >= 0.9.1 && <0.10
                , blaze-html     >= 0.8  && < 0.9
                , blaze-markup   >= 0.7  && < 0.8
                , bytestring

--- a/servant-auth-cookie.cabal
+++ b/servant-auth-cookie.cabal
@@ -1,5 +1,5 @@
 name:                servant-auth-cookie
-version:             0.3.2
+version:             0.4
 synopsis:            Authentication via encrypted cookies
 description:         Authentication via encrypted client-side cookies,
                      inspired by client-session library by Michael Snoyman and based on
@@ -22,6 +22,11 @@ flag dev
   manual:             True
   default:            False
 
+flag servant9
+  description:        Use servant-0.9
+  manual:             False
+  default:            True
+
 library
   exposed-modules:
     Servant.Server.Experimental.Auth.Cookie
@@ -38,8 +43,8 @@ library
                , http-types     >= 0.9   && < 0.10
                , memory         >= 0.11  && < 0.14
                , mtl            >= 2.0  && < 3.0
-               , servant        >= 0.5   && < 0.9
-               , servant-server >= 0.5   && < 0.9
+               , servant        >= 0.5   && < 0.10
+               , servant-server >= 0.5   && < 0.10
                , time           >= 1.5   && < 1.7
                , transformers   >= 0.4   && < 0.6
                , wai            >= 3.0   && < 3.3
@@ -52,7 +57,17 @@ library
   hs-source-dirs:      src
   default-language:    Haskell2010
 
-executable example
+  if flag(servant9)
+    build-depends:
+      servant >= 0.9,
+      http-api-data == 0.3.*
+  else
+    build-depends:
+      servant < 0.9,
+      bytestring-conversion >= 0.3.1 && <0.4
+
+test-suite example
+  type:                exitcode-stdio-1.0
   main-is:             Main.hs
   build-depends: base           >= 4.7  && < 5.0
                , blaze-html     >= 0.8  && < 0.9
@@ -63,10 +78,10 @@ executable example
                , data-default
                , http-media
                , mtl            >= 2.0  && < 3.0
-               , servant        >= 0.5  && < 0.9
+               , servant        >= 0.5  && < 0.10
                , servant-auth-cookie
-               , servant-blaze  >= 0.5  && < 0.9
-               , servant-server >= 0.5  && < 0.9
+               , servant-blaze  >= 0.5  && < 0.10
+               , servant-server >= 0.5  && < 0.10
                , text
                , wai            >= 3.0  && < 3.3
                , warp           >= 3.0  && < 3.3
@@ -76,6 +91,15 @@ executable example
     ghc-options:      -O2 -Wall
   hs-source-dirs:      example
   default-language:    Haskell2010
+
+  if flag(servant9)
+    build-depends:
+      servant >= 0.9,
+      http-api-data == 0.3.*
+  else
+    build-depends:
+      servant < 0.9,
+      bytestring-conversion >= 0.3.1 && <0.4
 
 test-suite tests
   type:           exitcode-stdio-1.0
@@ -95,7 +119,7 @@ test-suite tests
                , deepseq        >= 1.3  && < 1.5
                , hspec          >= 2.0  && < 3.0
                , servant-auth-cookie
-               , servant-server >= 0.5  && < 0.9
+               , servant-server >= 0.5  && < 0.10
                , time           >= 1.5  && < 1.7
   if !impl(ghc >= 7.8)
     build-depends:    tagged    == 0.8.*
@@ -114,7 +138,7 @@ benchmark bench
                , criterion      >= 0.6.2.1 && < 1.2
                , cryptonite     >= 0.14    && <= 0.20
                , servant-auth-cookie
-               , servant-server >= 0.5     && < 0.9
+               , servant-server >= 0.5     && < 0.10
   if flag(dev)
     ghc-options:      -Wall -Werror
   else

--- a/servant-auth-cookie.cabal
+++ b/servant-auth-cookie.cabal
@@ -45,11 +45,11 @@ library
                , mtl            >= 2.0  && < 3.0
                , servant        >= 0.5   && < 0.10
                , servant-server >= 0.5   && < 0.10
+               , tagged         == 0.8.*
                , time           >= 1.5   && < 1.7
                , transformers   >= 0.4   && < 0.6
                , wai            >= 3.0   && < 3.3
-  if !impl(ghc >= 7.8)
-    build-depends:    tagged    == 0.8.*
+
   if flag(dev)
     ghc-options:      -Wall -Werror
   else

--- a/src/Servant/Server/Experimental/Auth/Cookie.hs
+++ b/src/Servant/Server/Experimental/Auth/Cookie.hs
@@ -41,6 +41,9 @@ module Servant.Server.Experimental.Auth.Cookie
 
   , AuthCookieSettings (..)
 
+  , EncryptedSession (..)
+  , emptyEncryptedSession
+
   , encryptCookie
   , decryptCookie
 
@@ -96,6 +99,12 @@ import qualified Data.ByteString.Char8  as BSC8
 import Control.Applicative
 #endif
 
+#if MIN_VERSION_servant(0,9,0)
+import Servant (ToHttpApiData (..))
+#else
+import Data.ByteString.Conversion (ToByteString (..))
+#endif
+
 ----------------------------------------------------------------------------
 -- General types
 
@@ -112,6 +121,22 @@ data Cookie = Cookie
   , cookieExpirationTime :: UTCTime    -- ^ The cookie's expiration time
   , cookiePayload        :: ByteString -- ^ The payload
   } deriving (Eq, Show)
+
+-- | A newtype wrapper over 'ByteString'
+newtype EncryptedSession = EncryptedSession ByteString
+  deriving (Eq, Show, Typeable)
+
+emptyEncryptedSession :: EncryptedSession
+emptyEncryptedSession = EncryptedSession ""
+
+#if MIN_VERSION_servant(0,9,0)
+instance ToHttpApiData EncryptedSession where
+  toHeader (EncryptedSession s) = s
+  toUrlPiece = error "toUrlPiece @EncryptedSession: not implemented" 
+#else
+instance ToByteString EncryptedSession where
+  builder (EncryptedSession s) = builder s
+#endif
 
 -- | The exception is thrown when something goes wrong with this package.
 
@@ -386,7 +411,7 @@ addSession
   :: ( MonadIO m
      , MonadThrow m
      , Serialize a
-     , AddHeader (e :: Symbol) ByteString s r )
+     , AddHeader (e :: Symbol) EncryptedSession s r )
   => AuthCookieSettings -- ^ Options, see 'AuthCookieSettings'
   -> RandomSource     -- ^ Random source to use
   -> ServerKey         -- ^ 'ServerKey' to use
@@ -395,7 +420,7 @@ addSession
   -> m r               -- ^ Response with the session added
 addSession acs rs sk sessionData response = do
   header <- renderSession acs rs sk sessionData
-  return (addHeader header response)
+  return (addHeader (EncryptedSession header) response)
 
 -- | Add cookie session to error allowing to set cookie even if response is
 -- not 200.

--- a/src/Servant/Server/Experimental/Auth/Cookie.hs
+++ b/src/Servant/Server/Experimental/Auth/Cookie.hs
@@ -53,10 +53,10 @@ module Servant.Server.Experimental.Auth.Cookie
   , addSession
   , addSessionToErr
   , getSession
-  
+
   -- exposed for testing purpose
   , renderSession
-  
+
   , defaultAuthHandler
   ) where
 
@@ -79,6 +79,7 @@ import Data.Monoid ((<>))
 import Data.Proxy
 import Data.Serialize
 import Data.Time
+import Data.Tagged (Tagged (..), retag)
 import Data.Typeable
 import GHC.TypeLits (Symbol)
 import Network.HTTP.Types (hCookie)
@@ -132,7 +133,7 @@ emptyEncryptedSession = EncryptedSession ""
 #if MIN_VERSION_servant(0,9,0)
 instance ToHttpApiData EncryptedSession where
   toHeader (EncryptedSession s) = s
-  toUrlPiece = error "toUrlPiece @EncryptedSession: not implemented" 
+  toUrlPiece = error "toUrlPiece @EncryptedSession: not implemented"
 #else
 instance ToByteString EncryptedSession where
   builder (EncryptedSession s) = builder s
@@ -160,6 +161,23 @@ data AuthCookieException
   deriving (Eq, Show, Typeable)
 
 instance Exception AuthCookieException
+
+----------------------------------------------------------------------------
+-- Tags for various bytestrings
+
+-- | Tag encrypted cookie
+data EncryptedCookie
+
+-- | Tag for base64 serialized and encrypted cookie
+data SerializedEncryptedCookie
+
+base64Encode :: Tagged EncryptedCookie ByteString -> Tagged SerializedEncryptedCookie ByteString
+base64Encode = retag . fmap Base64.encode
+
+base64Decode
+  :: Tagged SerializedEncryptedCookie ByteString
+  -> Either String (Tagged EncryptedCookie ByteString)
+base64Decode = fmap Tagged . Base64.decode . unTagged
 
 ----------------------------------------------------------------------------
 -- Random source
@@ -211,7 +229,7 @@ mkServerKeyFromBytes :: MonadIO m
   => ByteString     -- ^ Predefined key
   -> m ServerKey    -- ^ New 'ServerKey'
 mkServerKeyFromBytes bytes =
-  ServerKey (BS.length bytes) Nothing `liftM` liftIO (newIORef (bytes, timeOrigin)) 
+  ServerKey (BS.length bytes) Nothing `liftM` liftIO (newIORef (bytes, timeOrigin))
   where
     -- we don't care about the time as the key never expires
     timeOrigin = UTCTime (toEnum 0) 0
@@ -298,7 +316,7 @@ encryptCookie :: (MonadIO m, MonadThrow m)
   => AuthCookieSettings -- ^ Options, see 'AuthCookieSettings'
   -> ServerKey         -- ^ 'ServerKey' to use
   -> Cookie            -- ^ The 'Cookie' to encrypt
-  -> m ByteString      -- ^ Encrypted 'Cookie' is form of 'ByteString'
+  -> m (Tagged EncryptedCookie ByteString)  -- ^ Encrypted 'Cookie' is form of 'ByteString'
 encryptCookie AuthCookieSettings {..} sk cookie = do
   let iv = cookieIV cookie
       expiration = BSC8.pack $ formatTime
@@ -313,7 +331,7 @@ encryptCookie AuthCookieSettings {..} sk cookie = do
     iv key (cookiePayload cookie)
   let mac = sign acsHashAlgorithm serverKey
         (BS.concat [iv, expiration, payload])
-  return . runPut $ do
+  return . Tagged . runPut $ do
     putByteString iv
     putByteString expiration
     putByteString payload
@@ -331,11 +349,11 @@ encryptCookie AuthCookieSettings {..} sk cookie = do
 --     * 'CannotParseExpirationTime'
 --     * 'CookieExpired'
 decryptCookie :: (MonadIO m, MonadThrow m)
-  => AuthCookieSettings -- ^ Options, see 'AuthCookieSettings'
-  -> ServerKey         -- ^ 'ServerKey' to use
-  -> ByteString        -- ^ The 'ByteString' to decrypt
-  -> m Cookie          -- ^ The decrypted 'Cookie'
-decryptCookie AuthCookieSettings {..} sk s = do
+  => AuthCookieSettings                 -- ^ Options, see 'AuthCookieSettings'
+  -> ServerKey                          -- ^ 'ServerKey' to use
+  -> Tagged EncryptedCookie ByteString  -- ^ The 'ByteString' to decrypt
+  -> m Cookie                           -- ^ The decrypted 'Cookie'
+decryptCookie AuthCookieSettings {..} sk (Tagged s) = do
   currentTime <- liftIO getCurrentTime
   let ivSize  = blockSize (unProxy acsCipher)
       expSize =
@@ -374,7 +392,7 @@ encryptSession :: (MonadIO m, MonadThrow m, Serialize a)
   -> RandomSource      -- ^ Random source to use
   -> ServerKey         -- ^ 'ServerKey' to use
   -> a                 -- ^ Session value
-  -> m ByteString     -- ^ Serialized and encrypted session
+  -> m (Tagged SerializedEncryptedCookie ByteString)  -- ^ Serialized and encrypted session
 encryptSession acs@AuthCookieSettings {..} randomSource sk session = do
   iv <- getRandomBytes randomSource (blockSize $ unProxy acsCipher)
   expirationTime <- liftM (addUTCTime acsMaxAge) (liftIO getCurrentTime)
@@ -384,7 +402,7 @@ encryptSession acs@AuthCookieSettings {..} randomSource sk session = do
         n  = BS.length payload
         l  = (bs - (n `rem` bs)) `rem` bs
     in getRandomBytes randomSource l
-  Base64.encode `liftM` encryptCookie acs sk (Cookie
+  base64Encode `liftM` encryptCookie acs sk (Cookie
     { cookieIV             = iv
     , cookieExpirationTime = expirationTime
     , cookiePayload        = BS.concat [payload, padding] })
@@ -392,13 +410,13 @@ encryptSession acs@AuthCookieSettings {..} randomSource sk session = do
 -- | Unpack session value from a cookie. The function can throw the same
 -- exceptions as 'decryptCookie'.
 decryptSession :: (MonadIO m, MonadThrow m, Serialize a)
-  => AuthCookieSettings -- ^ Options, see 'AuthCookieSettings'
-  -> ServerKey         -- ^ 'ServerKey' to use
-  -> ByteString        -- ^ Cookie in binary form
-  -> m a               -- ^ Unpacked session value
+  => AuthCookieSettings                           -- ^ Options, see 'AuthCookieSettings'
+  -> ServerKey                                    -- ^ 'ServerKey' to use
+  -> Tagged SerializedEncryptedCookie ByteString  -- ^ Cookie in binary form
+  -> m a                                          -- ^ Unpacked session value
 decryptSession acs@AuthCookieSettings {..} sk s =
   let fromRight = either (throwM . SessionDeserializationFailed) return
-  in fromRight (Base64.decode s) >>=
+  in fromRight (base64Decode s) >>=
      decryptCookie acs sk        >>=
      fromRight . runGet get . cookiePayload
 
@@ -450,7 +468,7 @@ getSession :: (MonadIO m, MonadThrow m, Serialize a)
 getSession acs@AuthCookieSettings {..} sk request = do
   let cookies = parseCookies <$> lookup hCookie (requestHeaders request)
       sessionBinary = cookies >>= lookup acsSessionField
-  maybe (return Nothing) (liftM Just . decryptSession acs sk) sessionBinary
+  maybe (return Nothing) (liftM Just . decryptSession acs sk . Tagged) sessionBinary
 
 -- | Render session cookie to 'ByteString'.
 renderSession
@@ -463,7 +481,7 @@ renderSession
   -> a
   -> m ByteString
 renderSession acs@AuthCookieSettings {..} rs sk sessionData = do
-  sessionBinary <- encryptSession acs rs sk sessionData
+  Tagged sessionBinary <- encryptSession acs rs sk sessionData
   let cookies =
         (acsSessionField, sessionBinary) :
         ("Path",    acsPath) :

--- a/tests/Main.hs
+++ b/tests/Main.hs
@@ -188,7 +188,7 @@ cipherId h c encryptAlgorithm decryptAlgorithm cookie encryptionHook = do
             , acsHashAlgorithm    = h
             , acsCipher           = c
             , .. }
-  encryptCookie sts sk cookie >>= decryptCookie sts sk . encryptionHook
+  encryptCookie sts sk cookie >>= decryptCookie sts sk . fmap encryptionHook
 
 sessionSpec :: Spec
 sessionSpec = do

